### DISCUSSION TITLE
hot reload not working inside codespaces

### DISCUSSION THREAD

**Discussion by fingers10** [community/community#16785] [Codespaces]
Dear Team,

Many thanks for enabling codespaces beta access for me. I'm exploring codespaces with my dotnet blazor wasm application from repo [I Love DotNet](https://github.com/ILoveDotNet/ilovedotnet). I'm able to run my application. I'm not able to use hot reload feature inside codespace. When I launch my application using `dotnet watch run` the app is up and running and when I make changes, the command is able to detect changes in file but hot reload is not happening. Any clues on this?

**Repro Steps:**

1. I opened the development branch of above repo in code space.
2. open terminal.
3. Run `cd ilovedotnet`
4. Run `cd Web`
5. Run `dotnet watch run`
6. Now app will launch in separate tab.
7. Now modify any file.
8. `dotnet watch run` command will detect file change and log message in terminal.
9. Now auto reload of app needs to happen which was opened in step 6.
10. But nothing happens.
11. When we press `ctrl + c` to terminate `dotnet watch run`.
12. You can notice hot reload related confirmation being asked after termination.

**Expected Behavior:**

after `dotnet watch run` any change to any file should trigger a auto reload and the app needs to reload in browser automatically.

**Here is the video recording of issue.**

https://user-images.githubusercontent.com/43729469/169107752-98d64282-843d-454d-a15d-ee550818fb97.mp4

 

Thanks,
Abdul Rahman

**Comment by anpaskin** [comment]
Hi Abdul,

Would you mind providing more detailed repro steps? We'd like to try this in a codespace and on local Mac and Windows machines and want to make sure we're running the same commands in the same locations.

Thanks!

**Reply by fingers10** [reply]
@anpaskin updated the question with repro steps.

**Comment by VincentDondain** [comment]
Hi @fingers10, I believe the error you're seeing with Hot Reload here is indeed a Codespaces bug (or rather a missing feature).

You can run that watch command with `--verbose` to get more info about that's going on and why hot reload could be failing.

In this case:

```
dotnet watch run --verbose
watch : Project supports hot reload and was configured to run with the default run-command. Watching with hot-reload
watch : Hot reload enabled. For a list of supported edits, see https://aka.ms/dotnet/hot-reload. Press "Ctrl + R" to restart.
watch : Building...
watch : Started '/home/codespace/.dotnet/dotnet' '' with process id 1536
  Determining projects to restore...
  All projects are up-to-date for restore.
  OOPSDemoComponents -> /workspaces/ilovedotnet/OOPSDemoComponents/bin/Debug/net6.0/OOPSDemoComponents.dll
  SharedModels -> /workspaces/ilovedotnet/SharedModels/bin/Debug/net6.0/SharedModels.dll
  BlazorDemoComponents -> /workspaces/ilovedotnet/BlazorDemoComponents/bin/Debug/net6.0/BlazorDemoComponents.dll
  Components -> /workspaces/ilovedotnet/Components/bin/Debug/net6.0/Components.dll
  Web -> /workspaces/ilovedotnet/Web/bin/Debug/net6.0/Web.dll
  Web (Blazor output) -> /workspaces/ilovedotnet/Web/bin/Debug/net6.0/wwwroot
watch : Process id 1536 ran for 4477ms
watch : Running MSBuild target 'GenerateWatchList' on '/workspaces/ilovedotnet/Web/Web.csproj'
watch : Started '/home/codespace/.dotnet/dotnet' '' with process id 1741
watch : Process id 1741 ran for 1172ms
watch : Watching 196 file(s) for changes
watch : dotnet-watch is configured to launch a browser on ASP.NET Core application startup.
watch : Configuring the app to use browser-refresh middleware.
watch : Refresh server running at wss://localhost:45711,ws://localhost:33781.
watch : HotReloadProfile: BlazorWebAssembly.
watch : Started 'dotnet' '"/var/nuget/microsoft.aspnetcore.components.webassembly.devserver/6.0.5/build/../tools/blazor-devserver.dll" --applicationpath "/workspaces/ilovedotnet/Web/bin/Debug/net6.0/Web.dll"' with process id 1823
watch : Running dotnet with the following arguments: run
watch : Started
watch : Waiting for a browser to connect
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://localhost:7176
watch : Launching browser.
watch : An exception occurred when attempting to launch a browser: System.ComponentModel.Win32Exception (2): An error occurred trying to start process 'https://localhost:7176/' with working directory '/workspaces/ilovedotnet/Web'. No such file or directory
   at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   at Microsoft.DotNet.Watcher.Tools.LaunchBrowserFilter.LaunchBrowser(String launchUrl)
   at Microsoft.DotNet.Watcher.Tools.LaunchBrowserFilter.OnOutput(Object sender, DataReceivedEventArgs eventArgs)
watch : Unable to launch the browser. Navigate to https://localhost:7176
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5176
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /workspaces/ilovedotnet/Web
```

It looks like the `dotnet watch` command wants to open a browser itself to bind to it and fails: `watch : Unable to launch the browser. Navigate to https://localhost:7176`, likely because it doesn't know about the forwarded port.

I opened an issue on our side to track this work and we'll report back here if/when we have a solution.
Please note that we cannot guarantee when this work will be prioritized.

**Reply by VincentDondain** [reply]
@fingers10 actually I just found a workaround that seems to work: try doing `export DOTNET_WATCH_SUPPRESS_LAUNCH_BROWSER="1"` in the Codespace.

That'll make it so the `dotnet watch` command doesn't launch the browser (which fails because it doesn't know about Port Forwarding), however you can open the forwarded port and then it seems that this "fixes" things where the browser with the forwarded port is then able to refresh when hotreload picks up changes.

**Reply by fingers10** [reply]
@VincentDondain, Please can you share the steps to export DOTNET_WATCH_SUPPRESS_LAUNCH_BROWSER="1" ?

**Reply by VincentDondain** [reply]
@fingers10 in the terminal of your Codespace, do `export DOTNET_WATCH_SUPPRESS_LAUNCH_BROWSER="1"`. You can then `echo  DOTNET_WATCH_SUPPRESS_LAUNCH_BROWSER` to confirm the environment variable has been set (:

**Reply by jomefavourite** [reply]
I'm having this same problem but with a React app, please is there a fix for it @VincentDondain

**Comment by jomefavourite** [comment]
The same issue but with a React app, please how do I go about it?

**Reply by FETXU** [reply]
I have found you can alter this line to your Package JSON to get hot reload in your package.json:

```
"scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  ```

to:

```
"scripts": {
    "start": "BROWSER=none WDS_SOCKET_PORT=0 react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  ```

**Reply by pascalfranke** [reply]
This worked for me. Thanks!

**Reply by Tejas-Naik** [reply]
Thanks, mate!

**Reply by HillelKott** [reply]
Thanks, save me a lot of refreshes

**Comment by TomlDev** [comment]
I had to add these lines to my webpack config:

```
const { merge } = require('webpack-merge');
const prodConfigCreator = require('./webpack.config');

module.exports = (config, context) => 
  merge(prodConfigCreator(config, context), {
    devServer: {
// THESE LINES WERE ADDED
      client: {
        webSocketURL: 'ws://0.0.0.0:0/ws',
      },
// UNTIL HERE WERE THE LINES ADDED
      proxy: {
        '/api': 'http://localhost:8081',
      },
    },
  });
